openmc/mgxs/library.py

```diff
@@ -0,0 +1,619 @@
+            raise ValueError(msg)
+
+        cv.check_type('coarse_groups', coarse_groups, openmc.mgxs.EnergyGroups)
+        cv.check_less_than('coarse groups', coarse_groups.num_groups,
+                           self.num_groups, equality=True)
+        cv.check_value('upper coarse energy', coarse_groups.group_edges[-1],
+                       [self.energy_groups.group_edges[-1]])
+        cv.check_value('lower coarse energy', coarse_groups.group_edges[0],
+                       [self.energy_groups.group_edges[0]])
+
+        # Clone this Library to initialize the condensed version
+        condensed_library = copy.deepcopy(self)
+        condensed_library.energy_groups = coarse_groups
+
+        # Condense the MGXS for each domain and mgxs type
+        for domain in self.domains:
+            for mgxs_type in self.mgxs_types:
+                mgxs = condensed_library.get_mgxs(domain, mgxs_type)
+                condensed_mgxs = mgxs.get_condensed_xs(coarse_groups)
+                condensed_library.all_mgxs[domain.id][mgxs_type] = condensed_mgxs
+
+        return condensed_library
+
+    def get_subdomain_avg_library(self):
+        """Construct a subdomain-averaged version of this library.
+
+        This routine averages each multi-group cross section across distribcell
+        instances. The method performs spatial homogenization to compute the
+        scalar flux-weighted average cross section across the subdomains.
+
+        NOTE: This method is only relevant for distribcell domain types and
+        simplys returns a deep copy of the library for all other domains types.
+
+        Returns
+        -------
+        Library
+            A new multi-group cross section library averaged across subdomains
+
+        Raises
+        ------
+        ValueError
+            When this method is called before a statepoint has been loaded
+
+        See also
+        --------
+        MGXS.get_subdomain_avg_xs(subdomains)
+
+        """
+
+        if self.statepoint is None:
+            msg = 'Unable to get a subdomain-averaged cross section ' \
+                  'library since the statepoint has not yet been loaded'
+            raise ValueError(msg)
+
+        # Clone this Library to initialize the subdomain-averaged version
+        subdomain_avg_library = copy.deepcopy(self)
+
+        if subdomain_avg_library.domain_type == 'distribcell':
+            subdomain_avg_library.domain_type = 'cell'
+        else:
+            return subdomain_avg_library
+
+        # Subdomain average the MGXS for each domain and mgxs type
+        for domain in self.domains:
+            for mgxs_type in self.mgxs_types:
+                mgxs = subdomain_avg_library.get_mgxs(domain, mgxs_type)
+                avg_mgxs = mgxs.get_subdomain_avg_xs()
+                subdomain_avg_library.all_mgxs[domain.id][mgxs_type] = avg_mgxs
+
+        return subdomain_avg_library
+
+    def build_hdf5_store(self, filename='mgxs', directory='mgxs',
```

Comment:
Why not just let the user specify the full path? Nothing inherently wrong if they want a .hdf5 extension or something else